lgc/builder/BuilderImpl.cpp

```diff
@@ -74,6 +74,161 @@ Value *BuilderImplBase::CreateDotProduct(Value *const vector1, Value *const vect
+  const bool hasNativeIntrinsic = isSupportCompBitwidth && isSupportSignedness && (isDot2 || isDot4 || isDot8);
+
+  // Create a lamda to clamp the input as expected
+  auto doClamp = [&](Value *input, Type *expectedTy) {
+    // Saturating
+    const unsigned resultWidth = expectedTy->getScalarSizeInBits();
+    // Compute the clamp range based on the
+    unsigned long long unsignedMax = (2ULL << (resultWidth - 1)) - 1;
+    long long signedMax = unsignedMax >> 1;
+    long long signedMin = -1LL - signedMax;
+
+    Value *minimum = nullptr, *maximum = nullptr;
+    Value *isUnderflow = nullptr, *isOverflow = nullptr;
+    if (isSigned) {
+      input = CreateSExt(input, getInt64Ty());
+      minimum = ConstantInt::getSigned(getInt64Ty(), signedMin);
+      maximum = ConstantInt::getSigned(getInt64Ty(), signedMax);
+      isUnderflow = CreateICmpSLT(input, minimum);
+      isOverflow = CreateICmpSGT(input, maximum);
+    } else {
+      input = CreateZExt(input, getInt64Ty());
+      minimum = getInt64(0);
+      maximum = getInt64(unsignedMax);
+      isUnderflow = CreateICmpULT(input, minimum);
+      isOverflow = CreateICmpUGT(input, maximum);
+    }
+    input = CreateSelect(isUnderflow, minimum, input);
+    input = CreateSelect(isOverflow, maximum, input);
+    return CreateTrunc(input, expectedTy);
+  };
+
+  auto input1 = vector1;
+  auto input2 = vector2;
+  Value *result = nullptr;
+  if (hasNativeIntrinsic) {
+    int supportedN = InvalidValue;
+    int intrinsic = InvalidValue;
+    if (isDot2) {
+      intrinsic = isBothSigned ? Intrinsic::amdgcn_sdot2 : Intrinsic::amdgcn_udot2;
+      supportedN = 2;
+    } else if (isDot4) {
+      intrinsic = isBothSigned ? Intrinsic::amdgcn_sdot4 : Intrinsic::amdgcn_udot4;
+      supportedN = 4;
+    } else {
+      assert(isDot8);
+      intrinsic = isBothSigned ? Intrinsic::amdgcn_sdot8 : Intrinsic::amdgcn_udot8;
+      supportedN = 8;
+    }
+    assert(intrinsic != InvalidValue);
+    // Do null-extension
+    SmallVector<int> shuffleMask;
+    for (int i = 0; i < supportedN; ++i)
+      shuffleMask.push_back(i);
+    input1 = CreateShuffleVector(input1, Constant::getNullValue(inputTy), shuffleMask);
+    input2 = CreateShuffleVector(input2, Constant::getNullValue(inputTy), shuffleMask);
+
+    // Cast to i32 for dot4 and dot8
+    if (compBitWidth == 4 || compBitWidth == 8) {
+      input1 = CreateBitCast(input1, getInt32Ty());
+      input2 = CreateBitCast(input2, getInt32Ty());
+    }
+
+    Value *clamp = hasAccumulator ? getTrue() : getFalse();
+    accumulator = isSigned ? CreateSExt(accumulator, getInt32Ty()) : CreateZExt(accumulator, getInt32Ty());
+    result = CreateIntrinsic(intrinsic, {}, {input1, input2, accumulator, clamp}, nullptr, instName);
+
+    if (resBitWidth < 32 && hasAccumulator)
+      result = doClamp(result, resultTy);
+    else
+      result = CreateTrunc(result, resultTy);
+  } else {
+    Value *sum = getInt32(0);
+    auto targetTy = getInt64Ty();
```

Comment:
Just use `resultTy` here, except when using the dot-based loop in which case it needs to be `i32`. Also, `sum` needs to be of `targetTy`.

Reply:
Will fix.